cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -1325,3 +1386,204 @@ PyTypeObject gpi_hdl_Object<gpi_cb_hdl>::py_type = []() -> PyTypeObject {
     type.tp_methods = gpi_cb_hdl_methods;
     return type;
 }();
+
+
+// Simulator clock
+// This provides support for a GPI-layer clock generator for increased performance.
+// The GPI callback used for driving the clock signal does not interact with Python
+// code and so does not suffer from the performance penalty
+
+struct sim_clock_data {
+  gpi_sim_hdl signal;
+  int64_t high_steps;
+  int64_t low_steps;
+  long next_value;
+  uint64_t  toggles;  // Support for cocotb.clock.Clock cycles
+  gpi_cb_hdl timed_callback;
+
+  // Jitter support
+  std::normal_distribution<float> *gaussian;
+  int64_t high_jitter;
+  int64_t low_jitter;
+};
+
+static PyObject *create_clock(PyObject *self, PyObject *args, PyObject *keywds)
+{
+  COCOTB_UNUSED(self);
+
+  if (!gpi_has_registered_impl()) {
+    PyErr_SetString(PyExc_RuntimeError, "No simulator available!");
+    return NULL;
+  }
+
+  PyObject *sig_hdl;
+  long long high_steps;
+  long long low_steps;
+  long long high_jitter = 0;
+  long long low_jitter = 0;
+  static const char *kwlist[] = {"signal", "high_steps", "low_steps",
+                                 "high_jitter", "low_jitter", NULL};
+
+  if (!PyArg_ParseTupleAndKeywords(args, keywds, "O!LL|LL", const_cast<char **>(kwlist),
+                                   &gpi_hdl_Object<gpi_sim_hdl>::py_type, &sig_hdl,
+                                   &high_steps, &low_steps, &high_jitter, &low_jitter)) {
+    return NULL;
+  }
+
+  if (high_steps < 0) {
+    PyErr_SetString(PyExc_ValueError, "high_steps value must be a non-negative integer");
+    return NULL;
+  }
+  if (low_steps < 0) {
+    PyErr_SetString(PyExc_ValueError, "low_steps value must be a non-negative integer");
+    return NULL;
+  }
+  if (high_jitter < 0) {
+    PyErr_SetString(PyExc_ValueError, "high_jitter value must be a non-negative integer");
+    return NULL;
+  }
+  if (low_jitter < 0) {
+    PyErr_SetString(PyExc_ValueError, "low_jitter value must be a non-negative integer");
+    return NULL;
+  }
+
+  sim_clock clock = new sim_clock_data{};
```

Comment:
This leaks. Is there anyway to tie it's lifetime to the lifetime of the Python object returned?

Reply:
See `sim_clock_dealloc()`.